core/src/main/java/kafka/server/share/DelayedShareFetch.java

```diff
@@ -119,18 +120,11 @@ public void onComplete() {
             });
 
             log.trace("Data successfully retrieved by replica manager: {}", responseData);
-            ShareFetchUtils.processFetchResponse(shareFetchData, responseData, partitionCacheMap, replicaManager)
-                .whenComplete((result, throwable) -> {
-                    if (throwable != null) {
-                        log.error("Error processing fetch response for share partitions", throwable);
-                        shareFetchData.future().completeExceptionally(throwable);
-                    } else {
-                        shareFetchData.future().complete(result);
-                    }
-                    // Releasing the lock to move ahead with the next request in queue.
-                    releasePartitionLocks(shareFetchData.groupId(), topicPartitionData.keySet());
-                });
-
+            Map<TopicIdPartition, ShareFetchResponseData.PartitionData> result =
```

Comment:
I think the semantics of the exception handling are different now. If processFetchResponse throws, you'll skip the release of the locks.

Reply:
I might be missing something here. There is `releasePartitionLocks` call as well in `catch` block for unknowns. Though I have moved removed them from both places and moved to finally block. Is that what you meant?

Reply:
Moving it to the finally block seems a bit odd because the lock isn’t set when `topicPartitionData` is empty.

Reply:
That's true but `releasePartitionLocks` iterates on passed `topicIdPartitions`. As they would be empty then no lock will be released, as expected.

But I think your comment is less related to behaviour but more related to that do we need to call the `release` if `topicIdPartitions` is empty, correct? Though I can add a check in `releasePartitionLocks` for emptiness of `topicIdPartitions` but that seemed to me redundant as anyways the forEach will not be iterated (also topicIdPartitions can never be null). Please let me know what you think.